keps/sig-network/2732-np-version/README.md

```diff
@@ -0,0 +1,803 @@
+newly-created NetworkPolicy is expected to be in effect, so that I don't
+time out the test case before the network plugin has managed to process
+the policy.
+
+#### Story 3 - Understanding Whether New Features Are Supported
+
+As a user, I want to be able to tell whether the cluster I am working
+in implements certain new NetworkPolicy features, such as
+[NetworkPolicy port ranges] or [selecting Namespaces by name].
+
+[NetworkPolicy port ranges]: https://github.com/kubernetes/enhancements/pull/2090
+[selecting Namespaces by name]: https://github.com/kubernetes/enhancements/pull/2113
+
+#### Story 4 - Reporting NetworkPolicy Version
+
+As a network plugin developer, I want to verify that a CNI implements a specific
+NetworkPolicy version.
+
+#### Story 5 - Changing an Alpha NetworkPolicy API
+
+As a kubernetes developer, I want to make a change to a previously-added
+alpha NetworkPolicy feature, in a way that does not cause problems for
+users in clusters with varying apiserver and network plugin versions.
+
+### Notes/Constraints/Caveats
+
+#### Feature Gates and NetworkPolicy
+
+Alpha APIs create an additional problem for NetworkPolicy, due to the
+three-way (user/apiserver/plugin) versioning split; if an alpha field
+is added in 1.21 but then changed in 1.22, then a 1.22 cluster with a
+network plugin implementing the 1.21 version would probably be unable
+to use either version of the feature. We might want to establish a
+somewhat more restrictive set of alpha API rules for NetworkPolicy
+changes than are used for other API types.
+
+#### Determining Whether a Network Plugin Implements NetworkPolicyStatus
+
+If a user is waiting to see the status of a newly-created NetworkPolicy,
+there is no entirely-reliable way to distinguish "the plugin has not yet
+set `status` but will soon" from "the plugin doesn't know about `status`
+and is never going to set it".
+
+It's not clear how big a problem this is, especially if we suggest that
+implementations should create an "empty" `status` right away if it's
+going to take them a while to determine the final `status`.
+
+#### Distributed and Delegating NetworkPolicy Implementations
+
+Some NetworkPolicy implementations do not operate from a single
+controller, but instead operate independently on every node (much like
+how kube-proxy implements Service rules). In this case it is not clear
+who would be responsible for updating `NetworkPolicyStatus`, and it may
+not be easy for any part of the system to know with certainty that a
+given NetworkPolicy was fully in effect across the cluster.
+
+In other cases, a plugin may implement NetworkPolicy in terms of some
+other lower-level policy engine (implemented by a cloud, perhaps), and
+may work by simply translating NetworkPolicy rules into rules for this
+other engine. As in the distributed case, it may not be possible for
+such a NetworkPolicy implementation to know for sure when a policy was
+fully in effect.
+
+### Risks and Mitigations
+
```

Comment:
Imagine that I perform a rolling replacement of all nodes, intending to apply what I think is a routine OS upgrade. After that upgrade, NetworkPolicy is no longer enforced. What happens to the `.status` of existing NetworkPolicies?

My thoughts on addressing this: maybe a heartbeat mechanism using Lease, that activates when the first Lease appears in a relevant namespace.

Reply:
@sftim That is a good question. We intentionally left out enforcing, so status without enforcing does seem strange. Maybe status should be in the future KEP that will deal with enforcing. See here for discussion: [network-policy-api-thread](https://kubernetes.slack.com/archives/C01CWSHQWQJ/p1624307942039800)